### ISSUE TITLE
SC2002 "Useless cat" with `wc` (cat needed to avoid printing filename)

### ISSUE THREAD

**Issue by refola** [koalaman/shellcheck#827]
I'm using `cat` to stop the `GNU coreutils 8.25` version of `wc` from printing the name of the file I'm counting lines of. ShellCheck calls this `cat` useless.

Rule Id: SC2002
I tried on shellcheck.net and verified that this is still a problem on the latest commit.

#### Snippet
```sh
#!/bin/sh
cat "$1" | wc -l
```

#### Here's what ShellCheck says:
```
Line 2:
cat "$1" | wc -l
    ^-- SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.
```

#### Here's what I wanted or expected to see:
I expected no output because the cat isn't useless. In particular, this use of `cat` seems to be the most direct way to prevent `wc` from echoing the name of the file.

**Comment by kurahaupo** [comment]
The simplest and most direct way to prevent `wc` from echoing the filename is simply
```
#!/bin/sh
wc -l <"$1"
```

This requires fewer processes, no inter-process pipe, and in theory allows `wc -c` to use `fstat()` to obtain the size of the file in bytes, rather than by reading its contents (although I'm unaware of any version that is implemented this way).

So yes, the warning is correct.

**Comment by kurahaupo** [comment]
Perhaps we need to explain better?

Change
```
cat foo | bar | zot | mumble
```
To
```
bar <foo | zot | mumble
```
Or if it's clearer to you
```
( bar | zot | mumble ) < foo
```
Or even
```
<foo bar | zot | mumble
```

**Comment by bittorf** [comment]
@refola please close if you agree

**Comment by refola** [comment]
Sorry for my late response.

I like the `<foo bar | zot | mumble` fix because it preserves the left-to-right visual flow of the data stream. But ShellCheck's original `cmd < file | ..` suggestion is still okay.

On January 21, 2017 10:55:30 PM MST, Martin Kealey <notifications@github.com> wrote:
>Perhaps we need to explain better?
>
>Change
>```
>cat foo | bar | zot | mumble
>```
>To
>```
>bar <foo | zot | mumble
>```
>Or if it's clearer to you
>```
>( bar | zot | mumble ) < foo
>```
>Or even
>```
><foo bar | zot | mumble
>```

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity and possibly replying from the 'wrong' email address.

**Comment by brchrisman** [comment]
This is also misleading when the argument to 'cat' is a variable which could intentionally be carrying multiple filenames and/or globbing. If that is the case, switching to wc -l $var will output filenames, and wc -l < $var will generate an error when 'var' is a whitespace delimited list of files.

**Comment by kurahaupo** [comment]
@brchrisman it's not a bug to show some false positives.

Warnings should be on by default, and can and should be disabled when you know that the code is safe.

Unless you're in strict posix mode, relying on wordsplitting is already discouraged. For the suggested use case it might be clearer to write like this:
```lang=sh
list=( $pattern )
cat "${list[@]}" | wc -l
```
